apis/Google.Cloud.Firestore/docs/datamodel.md

```diff
@@ -147,14 +147,99 @@ document data to a dictionary representation.
+value will be stored as a 64-bit integer in the same way as an `int`
+property in an attributed type, or an `int` value within a dictionary.
+
+The `FromFirestore` method receives the deserialized value using the
+default mapping, as shown in the earlier table.
+
+### Applying a converter to a type
+
+Once you've created a converter, you can either apply it to a type
+or an individual attributed property. In each case, the
+`ConverterType` property is specified in the attribute.
+
+For example, suppose you wish to create several ID classes or
+structs, each containing a string of the underlying ID. This is a
+technique sometimes used to effectively make type-safe identifiers.
+You could create a converter for each class, and apply the
+`[FirestoreData]` attribute to each class, specifying the
+corresponding converter. This would allow each ID to be stored as
+just the string, rather than as a map containing a single element with
+a string value. The following example demonstrates this with a
+`PlayerId` class. If another attributed class (e.g. a `Game`) had a
+`PlayerId` property, the converter would be used automatically.
+
+{{sample:DataModel.CustomConverterType}}
+
+### Applying a converter to a property
+
+Sometimes you may want to perform custom conversions for types that
+you can't apply attributes to, or you may want different conversions
+in different situations. In that case, you can implement the
+converter in the same way, but apply it selectively using the
+`CustomConverter` property on the `[FirestoreProperty]` attribute
+instead.
+
+As an example, you may want to use the .NET `Guid` struct within
+your data model. You could store each `Guid` property as a string,
+using the following sample code.
+
+{{sample:DataModel.CustomConverterProperty}}
+
+### Document converters
+
+If you wish a .NET type to be stored as a complete document after
+custom conversion, the converter must return a value which would be
+serialized as a map. This could be a dictionary, an anonymous type,
+or even a separate attributed type. When deserializing, the
+converter will receive an `IDictionary<string, object>` which it
+should use to extract the original data.
+
+{{sample:DataModel.ComplexConverter}}
+
+### Null values and custom converters
+
+Note that the conversion methods never receive null references, nor should
+they return null values. While it would be possible to deserialize
+null values to non-null .NET values, the conversion used in
+serialization is usually determined based on the actual type of the
+value being serialized. The approach used for null values ensures
+consistency.
+
+As a side-effect of this decision, it is advisable for the type
+argument of `IFirestoreConverter` to be a class or a non-nullable
+value type. For example, implement `IFirestoreConverter<Guid>`
+rather than `IFirestoreConverter<Guid?>`. If a converter is supplied
+for a non-nullable value type, the converter will automatically be
```

Comment:
the las part is confusing "the converter will automatically be used..." I would change it to, "the converter can be used...". Well, I suppose this has more to do with my previous comment on the automaticity of the application of converters in properties.

Reply:
I think this is covered by the previous bit.